libcore/src/SimulationHelper.cpp

```diff
@@ -0,0 +1,218 @@
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * any later version.
+ *
+ * JuPedSim is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with JuPedSim. If not, see <http://www.gnu.org/licenses/>.
+ **/
+#include "SimulationHelper.h"
+
+#include "general/Logger.h"
+#include "geometry/Room.h"
+#include "geometry/SubRoom.h"
+
+#include <algorithm>
+
+std::optional<bool> SimulationHelper::UpdateRoom(const Building & building, Pedestrian & ped)
+{
+    // No relocation needed, ped is in its assigned room/subroom
+    // TODO add check if room/subroom really exist
+    auto oldSubRoom = building.GetRoom(ped.GetRoomID())->GetSubRoom(ped.GetSubRoomID());
+    if(oldSubRoom->IsInSubRoom(&ped)) {
+        return false;
+    }
+
+    // relocation needed, look if pedestrian is in one of the neighboring rooms/subrooms
+    // get all connected directly connected rooms/subrooms
+    // TODO check for closed doors?
+    std::map<int, std::shared_ptr<SubRoom>> subroomsConnected;
+    for(const auto & [roomID, room] : building.GetAllRooms()) {
+        std::copy_if(
+            std::begin(room->GetAllSubRooms()),
+            std::end(room->GetAllSubRooms()),
+            std::inserter(subroomsConnected, subroomsConnected.end()),
+            [oldSubRoom](auto const & iterator) -> bool {
+                return iterator.second->IsDirectlyConnectedWith(oldSubRoom);
+            });
+    }
+
+    // check if pedestrian is in one of the neighboring rooms/subrooms
+    auto currentSubRoom = std::find_if(
+        std::begin(subroomsConnected),
+        std::end(subroomsConnected),
+        [&ped](auto const & iterator) -> bool {
+            return iterator.second->IsInSubRoom(ped.GetPos());
+        });
+
+    // pedestrian is in one of the neighboring rooms/subrooms
+    if(currentSubRoom != subroomsConnected.end()) {
+        ped.UpdateRoom(currentSubRoom->second->GetRoomID(), currentSubRoom->second->GetSubRoomID());
+        ped.SetSubRoomUID(currentSubRoom->second->GetUID());
+        return true;
+    } else {
+        ped.UpdateRoom(-1, -1);
+        ped.SetSubRoomUID(-1);
+        return std::nullopt;
+    }
+}
+
+std::vector<Pedestrian *> SimulationHelper::FindPedsReachedFinalGoal(
+    const Building & building,
+    const std::vector<Pedestrian *> & peds)
+{
+    std::vector<Pedestrian *> pedsAtFinalGoal;
+    const auto & goals = building.GetAllGoals();
+
+    std::copy_if(
+        std::begin(peds),
+        std::end(peds),
+        std::inserter(pedsAtFinalGoal, std::end(pedsAtFinalGoal)),
+        [goals](auto const ped) -> bool {
```

Comment:
```suggestion
        [goals](const Pedestrian * ped) -> bool {
```
I thinks auto is a bit missleading when you get a pointer. And we want the Pedestrian const and not the pointer.